internal/caveats/run.go

```diff
@@ -26,125 +31,101 @@ const (
 )
 
-// RunCaveatExpression runs a caveat expression over the given context and returns the result.
-func RunCaveatExpression(
+// RunSingleCaveatExpression runs a caveat expression over the given context and returns the result.
+// This instantiates its own CaveatRunner, and should therefore only be used in one-off situations.
+func RunSingleCaveatExpression(
 	ctx context.Context,
 	expr *core.CaveatExpression,
 	context map[string]any,
 	reader datastore.CaveatReader,
 	debugOption RunCaveatExpressionDebugOption,
 ) (ExpressionResult, error) {
-	env := caveats.NewEnvironment()
-	return runExpression(ctx, env, expr, context, reader, debugOption)
-}
-
-// ExpressionResult is the result of a caveat expression being run.
-type ExpressionResult interface {
-	// Value is the resolved value for the expression. For partially applied expressions, this value will be false.
-	Value() bool
-
-	// IsPartial returns whether the expression was only partially applied.
-	IsPartial() bool
-
-	// MissingVarNames returns the names of the parameters missing from the context.
-	MissingVarNames() ([]string, error)
+	runner := NewCaveatRunner()
+	return runner.RunCaveatExpression(ctx, expr, context, reader, debugOption)
 }
 
-type syntheticResult struct {
-	value           bool
-	isPartialResult bool
-
-	op                  core.CaveatOperation_Operation
-	exprResultsForDebug []ExpressionResult
-	missingVarNames     *mapz.Set[string]
-}
-
-func (sr syntheticResult) Value() bool {
-	return sr.value
-}
-
-func (sr syntheticResult) IsPartial() bool {
-	return sr.isPartialResult
+// CaveatRunner is a helper for running caveats, providing a cache for deserialized caveats.
+type CaveatRunner struct {
+	caveatDefs          map[string]*core.CaveatDefinition
+	deserializedCaveats map[string]*caveats.CompiledCaveat
 }
 
-func (sr syntheticResult) MissingVarNames() ([]string, error) {
-	if sr.isPartialResult {
-		if sr.missingVarNames != nil {
-			return sr.missingVarNames.AsSlice(), nil
-		}
-
-		missingVarNames := mapz.NewSet[string]()
-		for _, exprResult := range sr.exprResultsForDebug {
-			if exprResult.IsPartial() {
-				found, err := exprResult.MissingVarNames()
-				if err != nil {
-					return nil, err
-				}
-
-				missingVarNames.Extend(found)
-			}
-		}
-
-		return missingVarNames.AsSlice(), nil
+// NewCaveatRunner creates a new CaveatRunner.
+func NewCaveatRunner() *CaveatRunner {
+	return &CaveatRunner{
+		caveatDefs:          map[string]*core.CaveatDefinition{},
+		deserializedCaveats: map[string]*caveats.CompiledCaveat{},
 	}
-
-	return nil, fmt.Errorf("not a partial value")
-}
-
-func isFalseResult(result ExpressionResult) bool {
-	return !result.Value() && !result.IsPartial()
-}
-
-func isTrueResult(result ExpressionResult) bool {
-	return result.Value() && !result.IsPartial()
 }
 
-func runExpression(
+// RunCaveatExpression runs a caveat expression over the given context and returns the result.
+func (cr *CaveatRunner) RunCaveatExpression(
```

Comment:
It looks like we might be missing a unit test for this new struct and method.

Reply:
`RunSingleCaveatExpression` does invoke this method, but I'll add a unit test just the same

Reply:
Added